### PR TITLE
DAOS-4770 test: Add a Jenkinsfile timeout for run_test.

### PR DESCRIPTION
**Pull request by ashleypittman**
Set initial timeout to 20 minutes which should trigger a failure.

Skip-checkpatch: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
The timeout here worked to the extent that it stopped the Jenkins job and CI is no longer consuming any resources, it didn't however report failure back to Github.  @brianjmurrell is that acceptable for this ticket?  I would suggest a timeout of 1 hour would be appropriate, the job typically completes in 25-30 minutes.

**Comment by brianjmurrell** [comment]
Yes, this is because we do the GitHub status update inside the `runTest()` step and the `timeout()` step aborts that step hard, before it can be updated.

The solution to that would be to do the GitHub status update in the `post` block but we couldn't do that due to [JENKINS-39203](https://issues.jenkins-ci.org/browse/JENKINS-39203).  That issue has (allegedly) been since resolved, so we could potentially move the GitHub status update to `post` however the other reason we do the GitHub status update in the `runTest()` step is so that we can extract the test's status from the JUnit XML file and set the GitHub status based on that.  Jenkins doesn't do that for us.  Jenkins in fact doesn't provide any GitHub status at a per-stage granularity -- we do that ourselves.

So I think we are going to have to live with GitHub not getting any status update for timed-out jobs.  At least for now.

**Comment by ashleypittman** [comment]
> Yes, this is because we do the GitHub status update inside the `runTest()` step and the `timeout()` step aborts that step hard, before it can be updated.

It does however appear to call the always() step, although none of the success() or other steps get called.  If you're OK with this can you approve it and I'll find someone else to review.

I don't anticipate this happening very often, only if there's bugs in the node_local_test itself although I could be wrong.  We can always revisit this if it turns out that I'm wrong in this regard.

File: Jenkinsfile [hunk 1]
```diff
@@ -1068,43 +1068,45 @@ pipeline {
...
-                                           DAOS_BASE=${SL_PREFIX%/install*}
-                                           rm -f dnt.*.memcheck.xml vm_test.out
-                                           NODE=${NODELIST%%,*}
-                                           ssh $SSH_KEY_ARGS jenkins@$NODE "set -x
-                                               set -e
-                                               sudo bash -c 'echo \"1\" > /proc/sys/kernel/sysrq'
-                                               if grep /mnt/daos\\  /proc/mounts; then
-                                                   sudo umount /mnt/daos
-                                               else
+                        timeout(time:3, unit:'MINUTES') {
```
Comment (brianjmurrell) [inline Jenkinsfile:1071]: You probably want to update this:

```suggestion
                        timeout(time:60, unit:'MINUTES') {
```

**Comment by brianjmurrell** [comment]
> It does however appear to call the always() step,

Yes, it should still do the `post` block and any steps inside of it.  But for the run_test stage, there is only an `always` block.

> although none of the success() or other steps get called.

Because they don't actually exist.  They have a comment block wrapped around them, waiting for future use.

> If you're OK with this can you approve it and I'll find someone else to review.

It needs the timeout updated from 3 minutes to 60.

> I don't anticipate this happening very often, only if there's bugs in the node_local_test itself although I could be wrong. We can always revisit this if it turns out that I'm wrong in this regard.

I suspect you are right.  In any case, it should be pretty obvious what happened if it does trigger.

**Comment by ashleypittman** [comment]
> > It does however appear to call the always() step,
> 
> Yes, it should still do the `post` block and any steps inside of it. But for the run_test stage, there is only an `always` block.
> 
> > although none of the success() or other steps get called.
> 
> Because they don't actually exist. They have a comment block wrapped around them, waiting for future use.

Oh I see now, yes it's all done inside the runTest() function.  Of course.

> > If you're OK with this can you approve it and I'll find someone else to review.
> 
> It needs the timeout updated from 3 minutes to 60.

Ah yes, done.

**Comment by brianjmurrell** [comment]
Hrm.  A thought on how to achieve the GitHub status update in this case...

The timeout simply raises an exception and exceptions can be caught.  So if in `runTest()`, around say https://github.com/daos-stack/pipeline-lib/blob/606b8f2973c4f79d0f976293b66e025997f58d10/vars/runTest.groovy#L74-L75 we were to add a `try`/`catch()` block, we could still do the post-test processing in `runTest()` to post a status to GitHub.

**Comment by ashleypittman** [comment]
> Hrm. A thought on how to achieve the GitHub status update in this case...
> 
> The timeout simply raises an exception and exceptions can be caught. So if in `runTest()`, around say https://github.com/daos-stack/pipeline-lib/blob/606b8f2973c4f79d0f976293b66e025997f58d10/vars/runTest.groovy#L74-L75 we were to add a `try`/`catch()` block, we could still do the post-test processing in `runTest()` to post a status to GitHub.

That is a good idea, however my groovy is nowhere near good enough to attempt that, and I'm trying to make progress with this ticket to avoid a highly unlikely situation.  I'd be happy to file a bug or re-assign the current one to you if you want to expand on that approach.

**Comment by brianjmurrell** [comment]
> I'd be happy to file a bug

Yes please, do.  We can land this as is, accepting that the GitHub status will be left orphaned if this timeout has to be invoked and loop back around later to add `try`/`catch` behaviour in `runTest()` to handle that timeout and update the status accordingly.  Please reference this PR in the ticket for back-story.

**Pull request merged**